File: src/bmsx_cpp/emulator/devices/input_controller.cpp
#include "input_controller.h"

namespace bmsx {
namespace {

constexpr const char* INP_CONTEXT_ID = "inp_chip";

} // namespace

InputController::InputController(Memory& memory, Input& input, const StringPool& strings)
	: m_memory(memory)
	, m_input(input)
	, m_strings(strings)
	, m_defaultInputMapping(Input::getDefaultInputMapping()) {
}

void InputController::reset() {
	for (i32 playerIndex = 1; playerIndex <= PLAYERS_MAX; playerIndex += 1) {
		PlayerChipState& state = playerState(playerIndex);
		if (state.contextPushed) {
			m_input.getPlayerInput(playerIndex)->popContext(INP_CONTEXT_ID);
		}
		state.keyboard.clear();
		state.gamepad.clear();
		state.contextPushed = false;
	}
	m_memory.writeValue(IO_INP_PLAYER, valueNumber(1.0));
	m_memory.writeValue(IO_INP_STATUS, valueNumber(0.0));
	m_memory.writeValue(IO_INP_VALUE, valueNumber(0.0));
}

void InputController::onCtrlWrite() {
	switch (toU32(asNumber(m_memory.readValue(IO_INP_CTRL)))) {
		case INP_CTRL_COMMIT:
			commitAction();
			return;
		case INP_CTRL_ARM:
			return;
		case INP_CTRL_RESET:
			resetActions();
			return;
	}
}

void InputController::onQueryWrite() {
	const Value queryValue = m_memory.readValue(IO_INP_QUERY);
	const std::string& queryText = m_strings.toString(asStringId(queryValue));
	PlayerInput* const playerInput = m_input.getPlayerInput(currentPlayerIndex());
	const bool triggered = playerInput->checkActionTriggered(queryText);
	m_memory.writeValue(IO_INP_STATUS, valueNumber(triggered ? 1.0 : 0.0));
	m_memory.writeValue(IO_INP_VALUE, valueNumber(0.0));
}

void InputController::onConsumeWrite() {
	const std::string& actionNames = m_strings.toString(asStringId(m_memory.readValue(IO_INP_CONSUME)));
	PlayerInput* const playerInput = m_input.getPlayerInput(currentPlayerIndex());
	size_t start = 0;
	for (size_t index = 0; index <= actionNames.size(); index += 1) {
		if (index != actionNames.size() && actionNames[index] != ',') {
			continue;
		}
		playerInput->consumeAction(actionNames.substr(start, index - start));
		start = index + 1;
	}
}

InputController::PlayerChipState& InputController::playerState(i32 playerIndex) {
	return m_playerStates[static_cast<size_t>(playerIndex - 1)];
}

const InputController::PlayerChipState& InputController::playerState(i32 playerIndex) const {
	return m_playerStates[static_cast<size_t>(playerIndex - 1)];
}

i32 InputController::currentPlayerIndex() const {
	return static_cast<i32>(std::floor(asNumber(m_memory.readValue(IO_INP_PLAYER))));
}

void InputController::commitAction() {
	const i32 playerIndex = currentPlayerIndex();
	PlayerChipState& state = playerState(playerIndex);
	const std::string& actionName = m_strings.toString(asStringId(m_memory.readValue(IO_INP_ACTION)));
	const std::string& bindingsText = m_strings.toString(asStringId(m_memory.readValue(IO_INP_BIND)));
	std::vector<KeyboardBinding> keyboardBindings;
	std::vector<GamepadBinding> gamepadBindings;
	appendBindings(bindingsText, keyboardBindings, gamepadBindings);
	state.keyboard[actionName] = std::move(keyboardBindings);
	state.gamepad[actionName] = std::move(gamepadBindings);
	PlayerInput* const playerInput = m_input.getPlayerInput(playerIndex);
	if (state.contextPushed) {
		playerInput->popContext(INP_CONTEXT_ID);
	}
	InputMap map;
	map.keyboard = state.keyboard;
	map.gamepad = state.gamepad;
	playerInput->pushContext(INP_CONTEXT_ID, 100, map);
	state.contextPushed = true;
}

void InputController::resetActions() {
	const i32 playerIndex = currentPlayerIndex();
	PlayerChipState& state = playerState(playerIndex);
	if (state.contextPushed) {
		m_input.getPlayerInput(playerIndex)->popContext(INP_CONTEXT_ID);
	}
	state.keyboard.clear();
	state.gamepad.clear();
	state.contextPushed = false;
	m_memory.writeValue(IO_INP_STATUS, valueNumber(0.0));
	m_memory.writeValue(IO_INP_VALUE, valueNumber(0.0));
}

void InputController::appendBindings(
	const std::string& bindingsText,
	std::vector<KeyboardBinding>& keyboardBindings,
	std::vector<GamepadBinding>& gamepadBindings
) const {
	size_t start = 0;
	for (size_t index = 0; index <= bindingsText.size(); index += 1) {
		if (index != bindingsText.size() && bindingsText[index] != ',') {
			continue;
		}
		const std::string binding = bindingsText.substr(start, index - start);
		const auto keyboardIt = m_defaultInputMapping.keyboard.find(binding);
		if (keyboardIt != m_defaultInputMapping.keyboard.end()) {
			const std::vector<KeyboardBinding>& defaultBindings = keyboardIt->second;
			keyboardBindings.insert(keyboardBindings.end(), defaultBindings.begin(), defaultBindings.end());
		}
		gamepadBindings.push_back(GamepadBinding{ binding, std::nullopt });
		start = index + 1;
	}
}

} // namespace bmsx

File: src/bmsx/emulator/devices/input_controller.ts
import { Input } from '../../input/input';
import type { GamepadBinding, GamepadInputMapping, KeyboardBinding, KeyboardInputMapping } from '../../input/inputtypes';
import {
	INP_CTRL_COMMIT,
	INP_CTRL_LATCH,
	INP_CTRL_RESET,
	IO_INP_ACTION,
	IO_INP_BIND,
	IO_INP_CONSUME,
	IO_INP_CTRL,
	IO_INP_PLAYER,
	IO_INP_QUERY,
	IO_INP_STATUS,
	IO_INP_VALUE,
} from '../io';
import { Memory } from '../memory';
import type { StringValue } from '../string_pool';

const INP_CONTEXT_ID = 'inp_chip';

type PlayerChipState = {
	keyboard: KeyboardInputMapping;
	gamepad: GamepadInputMapping;
	contextPushed: boolean;
	latchedFrame: number;
};

export class InputController {
	private readonly playerStates: PlayerChipState[] = Array.from(
		{ length: Input.PLAYERS_MAX },
		(): PlayerChipState => ({
			keyboard: {},
			gamepad: {},
			contextPushed: false,
			latchedFrame: 0,
		}),
	);

	public constructor(
		private readonly memory: Memory,
		private readonly input: Input,
	) {}

	public reset(): void {
		for (let playerIndex = 1; playerIndex <= Input.PLAYERS_MAX; playerIndex += 1) {
			const state = this.playerStates[playerIndex - 1]!;
			if (state.contextPushed) {
				this.input.getPlayerInput(playerIndex).popContext(INP_CONTEXT_ID);
			}
			state.keyboard = {};
			state.gamepad = {};
			state.contextPushed = false;
			state.latchedFrame = 0;
		}
		this.memory.writeValue(IO_INP_PLAYER, 1);
		this.memory.writeValue(IO_INP_STATUS, 0);
		this.memory.writeValue(IO_INP_VALUE, 0);
	}

	public onCtrlWrite(): void {
		switch (this.memory.readValue(IO_INP_CTRL) as number) {
			case INP_CTRL_COMMIT:
				this.commitAction();
				return;
			case INP_CTRL_LATCH:
				this.latchInput();
				return;
			case INP_CTRL_RESET:
				this.resetActions();
				return;
		}
	}

	public onQueryWrite(): void {
		const query = this.memory.readValue(IO_INP_QUERY) as StringValue;
		const playerInput = this.input.getPlayerInput(this.memory.readValue(IO_INP_PLAYER) as number);
		const triggered = playerInput.checkActionTriggered(query.text);
		this.memory.writeValue(IO_INP_STATUS, triggered ? 1 : 0);
		this.memory.writeValue(IO_INP_VALUE, 0);
	}

	public onConsumeWrite(): void {
		const actionName = (this.memory.readValue(IO_INP_CONSUME) as StringValue).text;
		this.input.getPlayerInput(this.memory.readValue(IO_INP_PLAYER) as number).consumeAction(actionName);
	}

	private commitAction(): void {
		const playerIndex = this.memory.readValue(IO_INP_PLAYER) as number;
		const state = this.playerStates[playerIndex - 1]!;
		const actionName = (this.memory.readValue(IO_INP_ACTION) as StringValue).text;
		const bindingsText = (this.memory.readValue(IO_INP_BIND) as StringValue).text;
		const keyboardBindings: KeyboardBinding[] = [];
		const gamepadBindings: GamepadBinding[] = [];
		this.appendBindings(bindingsText, keyboardBindings, gamepadBindings);
		state.keyboard[actionName] = keyboardBindings;
		state.gamepad[actionName] = gamepadBindings;
		const playerInput = this.input.getPlayerInput(playerIndex);
		if (state.contextPushed) {
			playerInput.popContext(INP_CONTEXT_ID);
		}
		playerInput.pushContext(INP_CONTEXT_ID, state.keyboard, state.gamepad, {});
		state.contextPushed = true;
	}

	private latchInput(): void {
		const playerIndex = this.memory.readValue(IO_INP_PLAYER) as number;
		const playerInput = this.input.getPlayerInput(playerIndex);
		this.playerStates[playerIndex - 1]!.latchedFrame = playerInput.pollFrame;
	}

	private resetActions(): void {
		const playerIndex = this.memory.readValue(IO_INP_PLAYER) as number;
		const state = this.playerStates[playerIndex - 1]!;
		if (state.contextPushed) {
			this.input.getPlayerInput(playerIndex).popContext(INP_CONTEXT_ID);
		}
		state.keyboard = {};
		state.gamepad = {};
		state.contextPushed = false;
		state.latchedFrame = 0;
		this.memory.writeValue(IO_INP_STATUS, 0);
		this.memory.writeValue(IO_INP_VALUE, 0);
	}

	private appendBindings(bindingsText: string, keyboardBindings: KeyboardBinding[], gamepadBindings: GamepadBinding[]): void {
		let bindingStart = 0;
		for (let index = 0; index <= bindingsText.length; index += 1) {
			if (index !== bindingsText.length && bindingsText.charCodeAt(index) !== 44) {
				continue;
			}
			const binding = bindingsText.slice(bindingStart, index);
			const defaultKeyboardBindings = Input.DEFAULT_INPUT_MAPPING.keyboard[binding];
			if (defaultKeyboardBindings) {
				for (let bindingIndex = 0; bindingIndex < defaultKeyboardBindings.length; bindingIndex += 1) {
					keyboardBindings.push(defaultKeyboardBindings[bindingIndex]!);
				}
			}
			gamepadBindings.push({ id: binding as never });
			bindingStart = index + 1;
		}
	}
}
